staging/src/k8s.io/apiserver/plugin/pkg/audit/log/backend.go

```diff
@@ -34,60 +35,69 @@ const (
 	FormatJson = "json"
 )
 
+// The plugin name reported in error metrics.
+const pluginName = "log"
+
 // AllowedFormats are the formats known by log backend.
 var AllowedFormats = []string{
 	FormatLegacy,
 	FormatJson,
 }
 
-type backend struct {
+func NewBackend(out io.Writer, format string, groupVersion schema.GroupVersion) (audit.Backend, error) {
+	return newBlockingBackend(out, format, groupVersion), nil
+}
+
+type blockingBackend struct {
 	out          io.Writer
 	format       string
 	groupVersion schema.GroupVersion
+	mu           sync.Mutex
 }
 
-var _ audit.Backend = &backend{}
+var _ audit.Backend = &blockingBackend{}
 
-func NewBackend(out io.Writer, format string, groupVersion schema.GroupVersion) audit.Backend {
-	return &backend{
+func newBlockingBackend(out io.Writer, format string, groupVersion schema.GroupVersion) audit.Backend {
+	return &blockingBackend{
 		out:          out,
 		format:       format,
 		groupVersion: groupVersion,
+		mu:           sync.Mutex{},
 	}
 }
 
-func (b *backend) ProcessEvents(events ...*auditinternal.Event) {
+func (b *blockingBackend) ProcessEvents(events ...*auditinternal.Event) {
 	for _, ev := range events {
-		b.logEvent(ev)
-	}
-}
+		line := ""
+		switch b.format {
+		case FormatLegacy:
+			line = audit.EventString(ev) + "\n"
+		case FormatJson:
+			bs, err := runtime.Encode(audit.Codecs.LegacyCodec(b.groupVersion), ev)
+			if err != nil {
+				audit.HandlePluginError(pluginName, err, ev)
+				return
+			}
+			line = string(bs[:])
+		default:
+			audit.HandlePluginError(pluginName, fmt.Errorf("log format %q is not in list of known formats (%s)",
+				b.format, strings.Join(AllowedFormats, ",")), ev)
+			return
+		}
 
-func (b *backend) logEvent(ev *auditinternal.Event) {
-	line := ""
-	switch b.format {
-	case FormatLegacy:
-		line = audit.EventString(ev) + "\n"
-	case FormatJson:
-		bs, err := runtime.Encode(audit.Codecs.LegacyCodec(b.groupVersion), ev)
+		b.mu.Lock()
+		_, err := fmt.Fprint(b.out, line)
+		b.mu.Unlock()
```

Comment:
Not sure it's good enough to do lock acquire/release action for every single event.

Reply:
Because the Writer interface does not guarantee concurrent safety.

Reply:
If we use a buffered writer,  we may encounter overlap.

Reply:
I don't know which is better.
Can we use one new goroutine to do write operation?

@crassirostris What do you think?

Reply:
> If we use a buffered writer, we may encounter overlap.

That means, if we use unbuffered write, we should not acquire lock?

Reply:
Normally, we can not judge, unless we use the fixed writer.

Reply:
I think it's better to store all formatted lines and then under one lock write them all

Also, use defer for unlocking

Reply:
make sense